src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -248,6 +247,58 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
             return node;
         }
 
+        /// <summary>
+        /// Walks up the package dependency graph to check for cycle, potentially degraded package versions <see cref="DependencyResult"/>.
+        /// Cycle: A -> B -> A (cycle)
+        /// Downgrade: B depends up on D 1.0. Hence this method returns a downgrade while processing D 2.0 package. 
+        /// A -> B -> C -> D 2.0 (downgrade)
+        ///        -> D 1.0
+        /// </summary>
+        /// <param name="graphEdge">Graph Edge node to check for cycle or potential degrades</param>
+        /// <param name="dependency">Transitive package dependency</param>
+        /// <param name="rootpredicate">Func delegate to invoke when processing direct package dependency</param>
+        private static (DependencyResult dependencyResult, LibraryDependency conflictingDependency) WalkTreeCheckCycleAndNearestWins(GraphEdge<RemoteResolveResult> graphEdge,
+                        LibraryDependency dependency, Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> rootpredicate)
+        {
+            (DependencyResult, LibraryDependency) result = default;
+
+            var edge = graphEdge;
+            bool exitFlag = false;
+
+            while (edge != null && !exitFlag)
+            {
+                if (StringComparer.OrdinalIgnoreCase.Equals(edge.Item.Data.Match.Library.Name, dependency.Name))
+                {
+                    result = (DependencyResult.Cycle, null);
+                    exitFlag = true;
+                    break;
+                }
+
+                foreach (var d in edge.Item.Data.Dependencies)
+                {
+                    if (d != dependency && dependency.LibraryRange.IsEclipsedBy(d.LibraryRange))
+                    {
+                        if (d.LibraryRange.VersionRange != null &&
```

Comment:
A lot of this logic matches the recursive check, can you either change that directly or extra methods that can be used in both?

Reply:
Fixed in the latest commit. Please let me know the feedback.